File: Silver/2018-01/rental_jan2018.java
// Rental Service - USACO Silver January 2018 (http://www.usaco.org/index.php?page=viewproblem2&cpid=787)
// This problem was completed as homework for the USACO Silver 2 Class on 7/10/21.

import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class L19_rental {
    static class Store implements Comparable<Store> {
        int numGallons;
        int cost;

        Store(int numGallons, int cost) {
            this.numGallons = numGallons;
            this.cost = cost;
        }

        @Override
        public int compareTo(Store o) {
            return cost - o.cost;
        }
    }

    public static void main(String[] args) throws IOException {
        // input
        String problemName = "rental";
        BufferedReader br = new BufferedReader(new FileReader(problemName + ".in"));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int numCows = Integer.parseInt(st.nextToken());
        int numStores = Integer.parseInt(st.nextToken());
        int numRenters = Integer.parseInt(st.nextToken());
        int[] cows = new int[numCows];
        Store[] stores = new Store[numStores];
        int[] renters = new int[numRenters];
        for (int i=0; i<numCows; i++) {
            cows[i] = Integer.parseInt(br.readLine());
        }
        for (int i=0; i<numStores; i++) {
            st = new StringTokenizer(br.readLine());
            stores[i] = new Store(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        }
        for (int i=0; i<numRenters; i++) {
            renters[i] = Integer.parseInt(br.readLine());
        }

        // algorithm
        Arrays.sort(cows);
        Arrays.sort(stores);
        Arrays.sort(renters); // traverse all backwards

        // try to sell all cows to store (sell high cow first)
        int[] milkProfits = new int[numCows];
        int storePtr = numStores-1;
        long totalProfit = 0;

        for (int cowPtr=numCows-1; cowPtr>=0; cowPtr--) {
            int milkProfit = 0;
            int milkAmt = cows[cowPtr];

            while (storePtr >= 0 && milkAmt > 0) {
                if (milkAmt <= stores[storePtr].numGallons) {
                    // current store will buy all of my milk
                    milkProfit += stores[storePtr].cost * milkAmt;
                    stores[storePtr].numGallons -= milkAmt;

                    if (stores[storePtr].numGallons == 0) {
                        // store won't buy more milk
                        storePtr--;
                    }

                    break;
                } else {
                    // have more milk than store wants to buy
                    milkProfit += stores[storePtr].cost * stores[storePtr].numGallons;
                    stores[storePtr].numGallons = 0;
                    milkAmt -= stores[storePtr--].numGallons;
                }
            }

            milkProfits[cowPtr] = milkProfit;
            totalProfit += milkProfit;
        }

        // switch out all cows with renting (rent small cow first)
        long newProfit = totalProfit;
        long maxProfit = totalProfit;
        int cowPtr = 0;
        int rentPtr = numRenters-1;

        while (cowPtr < numCows && rentPtr >= 0) {
            newProfit -= milkProfits[cowPtr++];
            newProfit += renters[rentPtr--];

            maxProfit = Math.max(maxProfit, newProfit);
        }

        // output
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter(problemName + ".out")));

        pw.println(maxProfit);

        br.close();
        pw.close();
    }
}

File: Silver/2018-01/rental_jan2018.cpp
// Rental Service
// USACO Silver January 2018: http://www.usaco.org/index.php?page=viewproblem2&cpid=787

#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

struct Store {
    int qty;
    int price;

    bool operator<(const Store& other) {
        if (price == other.price) return qty > other.qty;
        return price > other.price;
    }
};

int n, m, r;
int cows[100001];
Store milk[100001];
int rental[100001];

int milkPrice[100001];

int main() {
    ifstream fin("rental.in");
    ofstream fout("rental.out");

    fin >> n >> m >> r;

    for (int i=1; i<=n; i++) fin >> cows[i];
    sort(cows+1, cows+n+1, greater<int>());

    for (int i=1; i<=m; i++) fin >> milk[i].qty >> milk[i].price;
    sort(milk+1, milk+m+1);

    for (int i=1; i<=r; i++) fin >> rental[i];
    sort(rental+1, rental+r+1, greater<int>());

    // calculate price for each cow's milk
    // milk of the cow with most milk is sold for highest price
    ll profit = 0;
    int storeIdx = 1;

    for (int i=1; i<=n; i++) {
        int amt = cows[i];

        while (amt > 0 && storeIdx <= m) {
            auto& [qty, price] = milk[storeIdx];

            int sellAmt = min(amt, qty);

            milkPrice[i] += sellAmt * price;
            qty -= sellAmt;
            amt -= sellAmt;

            if (qty == 0) storeIdx++;
        }

        profit += milkPrice[i];
        if (storeIdx > m) break;
    }

    ll result = profit;

    // rent cows with lowest milk for highest price
    // subtract their milk selling price and add rental price
    int cowIdx = n;
    int rentIdx = 1;

    while (rentIdx<=r && cowIdx>=1) {
        if (milkPrice[cowIdx] >= rental[rentIdx]) break;

        profit -= milkPrice[cowIdx--];
        profit += rental[rentIdx++];

        result = max(result, profit);
    }

    fout << result << endl;
}